### DISCUSSION TITLE
/posts/delegate-access-in-active-directory-with-powershell/

### DISCUSSION THREAD

**Discussion by giscus** [constantinhager/the-itguy#7] [Announcements]
# delegate-access-in-active-directory-with-powershell/

In this blog post I’m going to show you how to delegate Active Directory permissions to other Active Directory groups.

https://the-itguy.de/delegate-access-in-active-directory-with-powershell/

**Comment by mcdonamw** [comment]
Thank you for this!  One note, however.  It doesn't seem you ever actually use the two functions.  As such the variables $guidmap and $extendedright never actually gets set, thus subsequent code cannot run successfully.  Some may find it useful to correct this, especially newer folk.

**Reply by constantinhager** [reply]
Hi.
Good point. This was one of my earlier posts. 
I will correct that immediately.

**Reply by constantinhager** [reply]
Thank you again. I have corrected it.

**Reply by mcdonamw** [reply]
Awesome!  You're welcome, and thanks again.  Has been very helpful.

**Reply by constantinhager** [reply]
welcome. Glad I can help

**Comment by mcdonamw** [comment]
Stupid question... I'm showing the ActiveDirectoryAccessRule class constructor takes an IdentityReference, but you're passing a SecurityIdentifier object.  

I found similar code where they first create a SecurityIdentifier of the Group.Sid, but then create an IdentifyReference from that SecurityIdentifer and then pass the IdentityReference to the ActiveDirectoryAccessRule class.

Why is it you can just bypass the IdentityReference portion of this and just pass the SecurityIdentifier itself to the constructor?

Code I'm referencing:  https://learn.microsoft.com/en-us/answers/questions/438121/delegate-rights-to-a-ou-for-one-service-user-on-co.html?childToView=1123466#answer-1123466

**Reply by constantinhager** [reply]
Hi,
if you look at the .NET reference of the IdentityReference Class [Identity Reference Class](https://learn.microsoft.com/en-us/dotnet/api/system.security.principal.identityreference?view=net-7.0&viewFallbackFrom=dotnet-plat-ext-7.0)

it is derived from SecurityIdentifier. So you can use either an NTAccount like in the example you referenced or a SecurityIdentifier like I did.

**Reply by mcdonamw** [reply]
Forgot to reply originally, but that makes sense now.  Thanks!

**Comment by MacsInSpace** [comment]
Cheers for this article.
I had a colleague who lost access to what staff member had what laptop and needed to do an audit. 

While we could have just written to a CSV on a share, we wanted a better solution that could be replicated at other sites. 
The manual process was found here:
https://lockstepgroup.com/blog/fun-with-ad-custom-attributes/

In conjunction with extending the schema from here: 
https://www.linkedin.com/pulse/using-powershell-extend-user-schema-active-directory-james-sargent/

Using an updated version of the original script here: 
https://github.com/MacsInSpace/blog-scripts/blob/master/ActiveDirectory/Set-ComputerCustomAttributes.ps1

And adding the view in AD here:
https://github.com/MacsInSpace/blog-scripts/blob/master/ActiveDirectory/Add-ADUCAttribute.ps1

This blog has helped me complete the puzzle of automating it all to give "NT Authority\SELF" permission to update computer AD specific properties.

Big thanks!

**Reply by MacsInSpace** [reply]
Oops.Correction


And adding the view in AD here:
https://gallery.technet.microsoft.com/Extend-Active-Directory-ccad3d1a
- with special thanks to the wayback machine -

https://web.archive.org/web/20200328153227/https://gallery.technet.microsoft.com/Extend-Active-Directory-ccad3d1a

https://web.archive.org/web/20200416/https://gallery.technet.microsoft.com/Extend-Active-Directory-ccad3d1a/file/147298/2/Add-ADUCAttribute.ps1

**Reply by constantinhager** [reply]
glad I could help

**Comment by vdias** [comment]
Trying to clone some ACL from a user1 to a new user2, but i'm really having some problems on the "clone" part... could someone help...?

```
# Original user details
$ouDN = "OU=Test,OU=TestUser,DC=Company,DC=corp"
$username = "user1"

# New user details
$newUsername = "user2"

# Get the Organizational Unit
$ou = Get-ADOrganizationalUnit $ouDN

# Retrieve ACL for the Organizational Unit
$acl = Get-Acl -Path "AD:\$($ou.DistinguishedName)"

# Retrieve access rules for the original user
$userAccessRules = $acl.Access | Where-Object { $_.IdentityReference.Value -eq "company\$username" }

# Create a new access rule for the new user
$newUserAccessRule = New-Object System.DirectoryServices.ActiveDirectoryAccessRule(
					"company\$newUsername",
					$userAccessRules.ActiveDirectoryRights,
					$userAccessRules.AccessControlType,
					$userAccessRules.InheritanceType,
					$userAccessRules.PropagationFlags)

# Add the new access rule to the ACL
$acl.AddAccessRule($newUserAccessRule)

# Apply the modified ACL to the Organizational Unit
Set-Acl -Path "AD:\$($ou.DistinguishedName)" -AclObject $acl
```

**Comment by Curtmcgirt** [comment]
function New-ADDExtendedRightMap

you're missing the initial "<#" commenting out synopsis, description, example, notes

**Reply by constantinhager** [reply]
thank you corrected it
